// nuxi/dist/chunks/build.mjs
import { d as defineCommand, c as consola } from '../shared/nuxi.57fa43de.mjs';
import 'node:util';
import 'node:path';
import 'node:process';
import 'node:tty';
import { w as writeTypes } from '../shared/nuxi.3b5e40b8.mjs';
import { l as loadKit } from '../shared/nuxi.f65c0f6b.mjs';
import { a as clearBuildDir } from '../shared/nuxi.e016df24.mjs';
import { o as overrideEnv } from '../shared/nuxi.2509f57e.mjs';
import { s as showVersions } from '../shared/nuxi.aa87612b.mjs';
import { s as sharedArgs, l as legacyRootDirArgs } from '../shared/nuxi.01862219.mjs';
import { r as resolve, a as relative } from '../shared/nuxi.dcd5f40f.mjs';
import 'node:fs';
import 'node:perf_hooks';
import '../shared/nuxi.a685c563.mjs';
import './satisfies.mjs';
import '../shared/nuxi.357b3b4e.mjs';
import '../shared/nuxi.cc8dd4a9.mjs';
import 'node:url';
import '../shared/nuxi.615ae2d1.mjs';
import 'node:module';
import '../shared/nuxi.2ec68044.mjs';
import 'node:assert';
import 'node:v8';
import '../shared/nuxi.3da669df.mjs';
import 'crypto';
import 'fs';
import 'module';
import 'path';
import 'perf_hooks';
import 'os';
import 'vm';
import 'url';
import 'assert';
import 'process';
import 'v8';
import 'util';
import 'tty';
import 'stream';
import '../shared/nuxi.d1302898.mjs';
import 'events';
import 'node:stream';
import '../shared/nuxi.8d9be659.mjs';
import '../shared/nuxi.31a0124f.mjs';
import '../shared/nuxi.cb18516a.mjs';

const buildCommand = defineCommand({
  meta: {
    name: "build",
    description: "Build nuxt for production deployment"
  },
  args: {
    ...sharedArgs,
    prerender: {
      type: "boolean",
      description: "Build nuxt and prerender static routes"
    },
    dotenv: {
      type: "string",
      description: "Path to .env file"
    },
    ...legacyRootDirArgs
  },
  async run(ctx) {
    overrideEnv("production");
    const cwd = resolve(ctx.args.cwd || ctx.args.rootDir || ".");
    showVersions(cwd);
    const {
      loadNuxt,
      buildNuxt,
      useNitro,
      writeTypes: writeTypes$1 = writeTypes
    } = await loadKit(cwd);
    const nuxt = await loadNuxt({
      rootDir: cwd,
      dotenv: {
        cwd,
        fileName: ctx.args.dotenv
      },
      overrides: {
        logLevel: ctx.args.logLevel,
        // TODO: remove in 3.8
        _generate: ctx.args.prerender,
        ...ctx.args.prerender ? { nitro: { static: true } } : {},
        ...ctx.data?.overrides
      }
    });
    let nitro;
    try {
      nitro = useNitro?.();
    } catch {
    }
    await clearBuildDir(nuxt.options.buildDir);
    await writeTypes$1(nuxt);
    nuxt.hook("build:error", (err) => {
      consola.error("Nuxt Build Error:", err);
      process.exit(1);
    });
    await buildNuxt(nuxt);
    if (ctx.args.prerender) {
      if (!nuxt.options.ssr) {
        consola.warn(
          "HTML content not prerendered because `ssr: false` was set. You can read more in `https://nuxt.com/docs/getting-started/deployment#static-hosting`."
        );
      }
      const dir = nitro?.options.output.publicDir;
      const publicDir = dir ? relative(process.cwd(), dir) : ".output/public";
      consola.success(
        `You can now deploy \`${publicDir}\` to any static hosting!`
      );
    }
  }
});

export { buildCommand as default };
