// nuxi/dist/chunks/generate.mjs
import { d as defineCommand } from '../shared/nuxi.57fa43de.mjs';
import buildCommand from './build.mjs';
import { s as sharedArgs, l as legacyRootDirArgs } from '../shared/nuxi.01862219.mjs';
import 'node:util';
import 'node:path';
import 'node:process';
import 'node:tty';
import '../shared/nuxi.3b5e40b8.mjs';
import 'node:fs';
import 'node:perf_hooks';
import '../shared/nuxi.a685c563.mjs';
import './satisfies.mjs';
import '../shared/nuxi.357b3b4e.mjs';
import '../shared/nuxi.cc8dd4a9.mjs';
import 'node:url';
import '../shared/nuxi.615ae2d1.mjs';
import 'node:module';
import '../shared/nuxi.2ec68044.mjs';
import 'node:assert';
import 'node:v8';
import '../shared/nuxi.dcd5f40f.mjs';
import '../shared/nuxi.3da669df.mjs';
import 'crypto';
import 'fs';
import 'module';
import 'path';
import 'perf_hooks';
import 'os';
import 'vm';
import 'url';
import 'assert';
import 'process';
import 'v8';
import 'util';
import 'tty';
import 'stream';
import '../shared/nuxi.d1302898.mjs';
import 'events';
import 'node:stream';
import '../shared/nuxi.8d9be659.mjs';
import '../shared/nuxi.f65c0f6b.mjs';
import '../shared/nuxi.31a0124f.mjs';
import '../shared/nuxi.e016df24.mjs';
import '../shared/nuxi.2509f57e.mjs';
import '../shared/nuxi.aa87612b.mjs';
import '../shared/nuxi.cb18516a.mjs';

const generate = defineCommand({
  meta: {
    name: "generate",
    description: "Build Nuxt and prerender all routes"
  },
  args: {
    ...sharedArgs,
    ...legacyRootDirArgs,
    dotenv: {
      type: "string",
      description: "Path to .env file"
    }
  },
  async run(ctx) {
    ctx.args.prerender = true;
    await buildCommand.run(ctx);
  }
});

export { generate as default };
